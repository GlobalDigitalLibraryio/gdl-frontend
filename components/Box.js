// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 * 
 * See LICENSE
 */

import styled from 'styled-components';
import {
  fontSize,
  space,
  color,
  width,
  flex,
  responsiveStyle,
} from 'styled-system';
import tag from 'tag-hoc';
import height from './helpers/height';
import textAlign from './helpers/textAlign';

const order = responsiveStyle('order');

// The inline array here is the list of props that are blacklisted, meaning they aren't passed to the DOM node
const Base = tag([
  'width',
  'w',
  'm',
  'mt',
  'mr',
  'mb',
  'ml',
  'mx',
  'my',
  'p',
  'pt',
  'pr',
  'pb',
  'pl',
  'px',
  'py',
  'flex',
  'order',
  'wrap',
  'flexDirection',
  'align',
  'justify',
  'column',
  'h',
  'color',
  'bg',
  'textAlign',
])('div');

const Box = styled(Base)`
  ${fontSize} ${width} ${space} ${flex} ${order} ${height} ${color} ${textAlign};
`;

export default Box;

/* TODO: Figure out how to Flow type styled components
type responsiveProp = string | number | Array<string> | Array<number>; 
type responsiveStringProp = string | Array<string>;

type Props = {
  fontSize: responsiveProp,

  width: responsiveProp,
  w: responsiveProp,
  m: responsiveProp,
  mt: responsiveProp,
  mr: responsiveProp,
  mb: responsiveProp,
  ml: responsiveProp,
  mx: responsiveProp,
  my: responsiveProp,

  p: responsiveProp,
  pt: responsiveProp,
  pr: responsiveProp,
  pb: responsiveProp,
  pl: responsiveProp,
  px: responsiveProp,
  py: responsiveProp,

  order: responsiveProp,
  flex: responsiveStringProp,

  is: string | React.Component

  color: responsiveStringProp,
  bg: responsiveStringProp,
  align: responseStringProp,
} */
