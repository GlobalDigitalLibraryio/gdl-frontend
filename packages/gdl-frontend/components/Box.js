// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */

import styled from 'react-emotion';
import {
  fontSize,
  space,
  color,
  width,
  flex,
  textAlign,
  order
} from 'styled-system';
import height from '../style/height';

const Box = styled('div')(
  fontSize,
  width,
  space,
  flex,
  order,
  height,
  color,
  textAlign
);

export default Box;
