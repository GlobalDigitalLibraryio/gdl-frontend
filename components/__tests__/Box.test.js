// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */

import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createSerializer } from 'jest-emotion';
import * as emotion from 'emotion';
import Box from '../Box';

expect.addSnapshotSerializer(createSerializer(emotion));

test('Can have different paddings', () => {
  let tree = mount(<Box p={10} />);
  expect(toJson(tree)).toMatchSnapshot();

  tree = mount(<Box px={5} py={10} />);
  expect(toJson(tree)).toMatchSnapshot();

  tree = mount(<Box pb="1px" pl="2px" pt="3px" pr="4px" />);
  expect(toJson(tree)).toMatchSnapshot();
});

test('Can have different margins', () => {
  let tree = mount(<Box m={10} />);
  expect(toJson(tree)).toMatchSnapshot();

  tree = mount(<Box mx={5} my={10} />);
  expect(toJson(tree)).toMatchSnapshot();

  tree = mount(<Box mb="1px" ml="2px" mt="3px" mr="4px" />);
  expect(toJson(tree)).toMatchSnapshot();
});
