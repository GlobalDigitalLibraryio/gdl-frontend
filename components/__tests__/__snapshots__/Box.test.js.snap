// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can have different margins 1`] = `
.emotion-0 {
  margin: 10px;
}

<Styled(div)
  m={10}
>
  <div
    className="emotion-0"
  />
</Styled(div)>
`;

exports[`Can have different margins 2`] = `
.emotion-0 {
  margin-left: 5px;
  margin-right: 5px;
  margin-top: 10px;
  margin-bottom: 10px;
}

<Styled(div)
  mx={5}
  my={10}
>
  <div
    className="emotion-0"
  />
</Styled(div)>
`;

exports[`Can have different margins 3`] = `
.emotion-0 {
  margin-bottom: 1px;
  margin-left: 2px;
  margin-right: 4px;
  margin-top: 3px;
}

<Styled(div)
  mb="1px"
  ml="2px"
  mr="4px"
  mt="3px"
>
  <div
    className="emotion-0"
  />
</Styled(div)>
`;

exports[`Can have different paddings 1`] = `
.emotion-0 {
  padding: 10px;
}

<Styled(div)
  p={10}
>
  <div
    className="emotion-0"
  />
</Styled(div)>
`;

exports[`Can have different paddings 2`] = `
.emotion-0 {
  padding-left: 5px;
  padding-right: 5px;
  padding-top: 10px;
  padding-bottom: 10px;
}

<Styled(div)
  px={5}
  py={10}
>
  <div
    className="emotion-0"
  />
</Styled(div)>
`;

exports[`Can have different paddings 3`] = `
.emotion-0 {
  padding-bottom: 1px;
  padding-left: 2px;
  padding-right: 4px;
  padding-top: 3px;
}

<Styled(div)
  pb="1px"
  pl="2px"
  pr="4px"
  pt="3px"
>
  <div
    className="emotion-0"
  />
</Styled(div)>
`;
